### DISCUSSION TITLE
Data provider: read related table fields with external mysql db

### DISCUSSION THREAD

**Discussion by HVStechnik** [Budibase/budibase#8233] [Help] [answered]
Dear Community

I very much apprechiate the possiblity to filter based on fields in related tables in sql databases. But, I can't manage to retrieve data with related information. Is there any option to do this without a second data provider / query to the db?

**Example**:
I have 2 tables in a mysql db: teachers and pupils with a 1:n relationship: one teacher, multiple pupils.
I would like to list all pupils with the age of their teacher. Age is a field within the teachers table.
I have a data provider, retrieving the pupils with some filters. And it great that we now can even filter for the teachers age. 
However, in the table I can not display the teachers age easily. 

The only option I found is to add another data provider to the table, which looks up the teacher with the teacher id from the pupils table. Then, you have to insert a repeater inside the data provider and put a pragraph inside, which displays the teachers age. 
For 50 pupils on a page, this results in 50 queries to get the teachers' age.

Is there any other option to achieve this?

Many thanks!

**Comment by melohagan** [comment, marked as answer]
Hey @HVStechnik 

Have you explored custom queries? https://docs.budibase.com/docs/custom-queries

**Reply by HVStechnik** [reply]
Hi @melohagan 

Sure. I'm currently working with custom queries. Yet, custom queries have the significant disadvantage that all fontend filtering and pagingation is not handled server-side unless we code that manually into the custom query, which will get very messy quickly. See #8214

I just thought: Great, we can filter for fields in related tables! But wait, why isn't it possbile to display the column, for which I was just able to create a filter... 
This seems to be a quick win, if it's not availabe somewhere.

**Reply by melohagan** [reply]
Would formulas may help you here? https://docs.budibase.com/docs/formula#fields-from-relationships

**Reply by HVStechnik** [reply]
This is working! Thank you! Yet, I think that it would be much easier, if we could have the same "relationship browser" as in the formula field binding in the data provider. 
My suggestion is that the data provider component should have an option, which relationships from the main table should be retieved. Then, in the contained tables and binding could use the related fields. 
Does it make sense to create a feature request for this?

Further question: While thinking about this, I was wondering whether Budibase does a db join in the data provider request to get the `display column` of a related table only if the display column is acutally needed in the UI or whether this is done independently of whether the relation is acutally displayed? 
From a computational perspective, the first would make much more sense. Then, we could also offer all related fields of the table retrieved by the data provider - similar as in the formula field binding - and build the query for the data provider dynamically for the component based on which columns are required by the UI.

**Reply by melohagan** [reply]
The formula fields are calculated as part of the backend query, which will happen regardless of whether it is being used in an app UI table.

Formulas in general are a little bit strange, but I'm glad it worked out for you! In my opinion, they wouldn't be needed if the internal views were more flexible and/or there were more query options for the internal db.
I think the team is thinking about ways to improve the internal db experience but there's nothing set in stone yet. 

But in your case being MySQL, it's probably a case of making the custom queries more powerful and integrated into the UI tables, e.g. pagination, etc.
